cloudnew/ec2.go

```diff
@@ -0,0 +1,518 @@
+	return hostutil.CheckSSHResponse(context.TODO(), h, opts)
+}
+
+// GetDNSName returns the DNS name for the host.
+func (m *ec2Manager) GetDNSName(h *host.Host) (string, error) {
+	info, err := m.getInstanceInfo(h.Id)
+	if err != nil {
+		return "", err
+	}
+	return *info.PublicDnsName, nil
+}
+
+// GetSSHOptions returns the command-line args to pass to SSH.
+func (m *ec2Manager) GetSSHOptions(h *host.Host, keyName string) ([]string, error) {
+	if keyName == "" {
+		return nil, errors.New("No key specified for EC2 host")
+	}
+	opts := []string{"-i", keyName}
+	hasKnownHostsFile := false
+
+	for _, opt := range h.Distro.SSHOptions {
+		opt = strings.Trim(opt, " \t")
+		opts = append(opts, "-o", opt)
+		if strings.HasPrefix(opt, "UserKnownHostsFile") {
+			hasKnownHostsFile = true
+		}
+	}
+
+	if !hasKnownHostsFile {
+		opts = append(opts, "-o", "UserKnownHostsFile=/dev/null")
+	}
+	return opts, nil
+}
+
+// TimeTilNextPayment returns how long until the next payment is due for a host.
+func (m *ec2Manager) TimeTilNextPayment(host *host.Host) time.Duration {
+	return timeTilNextEC2Payment(host)
+}
+
+// GetInstanceName returns the name of an instance.
+func (m *ec2Manager) GetInstanceName(d *distro.Distro) string {
+	return d.GenerateName()
+}
+
+// makeInt64Ptr is necessary because Go does not allow you to write `&int64(1)`.
+func makeInt64Ptr(i int64) *int64 {
+	return &i
+}
+
+// makeStringPtr is necessary because Go does not allow you to write `&"foo"`.
+func makeStringPtr(s string) *string {
+	return &s
+}
+
+func (m *ec2Manager) getInstanceInfo(id string) (*ec2.Instance, error) {
+	resp, err := m.client.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{makeStringPtr(id)},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "EC2 API returns error for DescribeInstances")
+	}
+	reservation := resp.Reservations
+	if len(reservation) == 0 {
+		err = errors.Errorf("No reservation found for instance id: %s", id)
+		grip.Error(err)
+		return nil, err
+	}
+
+	instances := reservation[0].Instances
+	if len(instances) == 0 {
+		err = errors.Errorf("'%s' was not found in reservation '%s'",
+			id, *resp.Reservations[0].ReservationId)
+		grip.Error(err)
+		return nil, err
+	}
+
+	return instances[0], nil
+}
+
+func (m *ec2Manager) getSpotInstanceStatus(id string) (cloud.CloudStatus, error) {
+	spotDetails, err := m.client.DescribeSpotInstanceRequests(&ec2.DescribeSpotInstanceRequestsInput{
+		SpotInstanceRequestIds: []*string{makeStringPtr(id)},
+	})
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get spot request info for %s", id)
+		grip.Error(err)
+		return cloud.StatusUnknown, err
+	}
+
+	spotInstance := spotDetails.SpotInstanceRequests[0]
+	//Spot request has been fulfilled, so get status of the instance itself
+	if *spotInstance.InstanceId != "" {
+		instanceInfo, err := m.getInstanceInfo(*spotInstance.InstanceId)
+		if err != nil {
+			err = errors.Wrap(err, "Got an error checking spot details")
+			grip.Error(err)
```

Comment:
consider using message.WrapError(err, message.Fields{}) to annotate this message with instance IDs/etc.